frontends/common/constantFolding.cpp

```diff
@@ -802,17 +802,22 @@ const IR::Node *DoConstantFolding::shift(const IR::Operation_Binary *e) {
     unsigned shift = static_cast<unsigned>(shift_amt->asInt());
     if (overflowWidth(e, shift)) return e;
 
-    auto tb = left->type->to<IR::Type_Bits>();
-    if (tb != nullptr) {
-        if (((unsigned)tb->width_bits() <= shift) && warnings)
-            ::warning(ErrorType::WARN_OVERFLOW, "%1%: Shifting %2%-bit value with %3%", e,
-                      tb->width_bits(), shift);
-    }
-
-    if (e->is<IR::Shl>())
+    if (e->is<IR::Shl>()) {
         value = Util::shift_left(value, shift);
-    else
+    } else {
         value = Util::shift_right(value, shift);
+    }
+    if (const auto *tb = left->type->to<IR::Type_Bits>()) {
+        // According to the P4 specification, the result of a shift that is larger than the width of
+        // the bit type is 0.
```

Comment:
unless it is a negative signed value being shifted right, in which case it should be -1 (sign bit shifted into all bit positions)